### ISSUE TITLE
Idea: Add new commands to the plugin (data import/export, getting debug logs...)

### ISSUE THREAD

**Issue by ognjen-vuceljic** [xixiaofinland/sf.nvim#127]
Hi @xixiaofinland ,

hope you're doing well today.

Just an idea - it might be good to consider adding more options to the plugin that can be triggered via hotkeys, in the same way as current ones do.

The commands that came to mind:

`sf data import/export`

[https://github.com/forcedotcom/cli/issues/2738](https://github.com/forcedotcom/cli/issues/2738)

Sounds like Salesforce has invested some resources in improving those commands, so we now can pull more than 2 levels of child object data, plus handle >200 records (they will be split into batches), and much more as described in the link above.  

Data export/import might be useful when we need to ensure that necessary data needs to be passed over from one environment to another.

------------------------------------

`sf apex get log`

There could be some interesting use cases here in [sf apex get log](https://developer.salesforce.com/docs/atlas.en-us.sfdx_cli_reference.meta/sfdx_cli_reference/cli_reference_apex_commands_unified.htm#cli_reference_apex_get_log_unified) command.

For example, we could fetch the latest generated log from the org by running `sf apex get log --number 1 -o  target_org`, or pass the `--log-id` as parameter. 

We could easily run those commands in terminal outside of plugin, but just throwing out an idea here for brainstorming, to see if this is something that makes sense, and is worth investing time in.

------------------------------------

Bonus question (nice to have):

- some sf commands, such as `sf apex get log` command, generate logs that are color-coded, which is quite useful for visualizing results

![image](https://github.com/user-attachments/assets/2e8a4105-e182-42e0-bb79-acc48310f9ac)

On the other hand, running anonymous apex comes with the same result structure, but no highlighting

![image](https://github.com/user-attachments/assets/0a409d36-89c4-484b-a94e-92eb8a2491c2)

Is this something we can standardize, so that printed commands are color coded (e.g. they have the filetype set to `log`, or something that would have syntax highlighting enabled).

Thanks!

**Comment by xixiaofinland** [comment]
Hi @ognjen-vuceljic, 

For those commands, they can be easily defined by users separately as described [here](https://github.com/xixiaofinland/sf.nvim#-custom-hotkeys).  It's hard for me to decide what should be part of the plugin or not. Maybe when I have the need or the plugin has more log-related features, I can add them into the plugin.

Speaking of the colored Salesforce log, yes, in case users retrieve and view them locally, it makes sense to have syntax highlight.

It might be a good idea to add a tree-sitter parser for Saleforce log, and give it a file type in this plugin. 
Alternatively, we can use reg-ex to parse since the text pattern is relatively simple.  
I'm more in favor of creating a tree-sitter parser, much more interesting :). But it will take some dedicated time.

Thanks again for sharing all your thoughts and the bug feedback. I won't have the drive to implement the sign icon feature without your support. I added you into the [contributor list](https://github.com/xixiaofinland/sf.nvim#-contributors) 😄 hope you don't mind. (Because markdown doesn't show github default profile icon properly, you don't have it there....

**Comment by xixiaofinland** [comment]
@ognjen-vuceljic 

How do you use the log locally? More context on this LinkedIn Post: https://www.linkedin.com/posts/xixiaofinland_salesforce-activity-7224647096610492416-hBEN?utm_source=share&utm_medium=member_desktop

**Comment by ognjen-vuceljic** [comment]
Thanks @xixiaofinland 

I usually download the log locally, and use one of 2 things in VS Code:

- Replay Debugger (which gives me a proper "Debug" environment, with breakpoints, etc.)
- I open the log, and use another extension called "Text Power Tools" which gives me the quick, nice filter ability, so I can easily focus on debug logs, or entry statements, or any line which contains a particular keyword


One thing I think would be extremely useful is highlighting the scope of a particular method inside log.

For example, pairing `METHOD_ENTRY` line with `METHOD_EXIT` line somehow, so that we understand when a particular method execution started, and when it finished.

Please see the example below:

![image](https://github.com/user-attachments/assets/ecf80e77-6c27-45ed-8ccf-8bb31f54677c)

This is a simple example, but gives us the ability to understand when the method started, what happened inside, and when it was completed. 

The tricky part is when one method calls another, which adds call invocations into a stack.


I think parsing the log as TXT file with some sort of fuzzy finding and searching would be a great start. For example, searching for lines that contain `|DEBUG|` inside would filter out only the lines with that search param, and leave out the rest.

Hope this makes sense as a starting point :)

**Comment by ognjen-vuceljic** [comment]
@xixiaofinland  I like the idea of `sf apex tail log | grep 'hello`

Could you, please, share some examples how you use it?

I do structure my debug statements in a way that clear ticket/JIRA identified is added, which makes search easier - but the issue is that those statements are scattered among multiple different log files, which can be quite big when we have queueables, future handlers, etc.

Any examples of usage would be much appreciated :)

**Comment by xixiaofinland** [comment]
This can be easily answered by GPT already below.

You simply insert `System.debug('hello:' + variables);` in the code and execute. the command will display the result on the fly.

```
The command sf apex tail log | grep 'hello' is used to monitor Salesforce Apex logs in real-time and filter them for lines containing the word "hello". Here’s a breakdown of what each part does:

sf apex tail log: This part of the command uses the Salesforce CLI (sf) to stream the Apex debug logs in real-time.
|: The pipe symbol (|) is used to take the output from the previous command (sf apex tail log) and pass it as input to the next command.
grep 'hello': This part of the command filters the streamed log output to show only the lines that contain the word "hello".
In summary, this command is useful for developers who want to monitor live Apex logs from Salesforce and focus on specific log entries containing the keyword "hello".
```

**Comment by xixiaofinland** [comment]
Thanks for sharing. Yes, once the parser is ready, this will be automatically supported.

> I think parsing the log as TXT file with some sort of fuzzy finding and searching would be a great start. For example, searching for lines that contain `|DEBUG|` inside would filter out only the lines with that search param, and leave out the rest.
>

**Comment by xixiaofinland** [comment]
Nvim-treesitter now has `sflog` parser merged. https://github.com/nvim-treesitter/nvim-treesitter/pull/7018#event-13827323464 

Update nvim-treesitter plugin to latest and add `sflog` parser like [here](https://github.com/xixiaofinland/dotfiles-nix/blob/55081dd2394030cc418778b311ba3fd7fb3ff6c8/dotfiles/nvim_config/lua/plugins/nvim-tree-sitter.lua#L28)

I will add this filetype into `sf.nvim` soon @ognjen-vuceljic

**Comment by xixiaofinland** [comment]
![Image 009](https://github.com/user-attachments/assets/bf938313-8f95-4fdd-9a57-783e2b8b8fdc)
